### ISSUE TITLE
Support pen input prediction and compensation

### ISSUE THREAD

**Issue by Snowmander** [ppy/osu-framework#2817]
Unlike Wacom pen tablets like CTL-480, modern styluses such as Apple Pencil and Surface Pen rely on prediction to improve responsiveness. On drawing apps, pen strokes are drawn based on the OS-provided prediction first, and then are corrected after the true position is available.

I suggest that the pen input prediction on osu should work like this:
1. The game should initially use the prediction to determine 300, 100, 50, or miss.
2. When the true position is available and the outcome conflicts, the visual effects and hitsounds of the note should be cancelled immediately and be replaced. (For example, if the true input was available 90ms later -> the visual effects / hitsounds play as if they had begun 90ms before).
3. The game corrects the scoring.

Unfortunately, although the APIs provide the correct past position of the pen, they do not seem to provide how much time has passed since. If some leniency is okay, the game could try to calculate the timing of the past pen position and use that to calculate scores instead, which will achieve latency-less input at a very small cost of inaccuracy.

Alternatively, the game could feature its own input prediction algorithm.

I understand that this is a big task that not many people might be interested in (and I don’t have the time right now to be better at programming right now), but I think this is going to be a vital part of the game for non-desktops.

Apple pencil:
https://developer.apple.com/documentation/uikit/touches_presses_and_gestures/handling_touches_in_your_view/minimizing_latency_with_predicted_touches

**Comment by peppy** [comment]
The osu! implementation you mention is overly complex. Prediction should either be be 100% used for positional input or 100% ignored. I think using may be best, but will require some real-world testing.

**Comment by UselessToucan** [comment]
> I think this is going to be a vital part of the game for non-desktops.

Can you provide any concrete use cases of a stylus? I have a Surface Pro 2 tablet with a Surface Pen stylus and I find it useless for gaming.

**Comment by Snowmander** [comment]
@UselessToucan My iPad Pro 3rd gen. I think it has the potential to replace desktop for osu! if pen support is better and touchscreen z/x input is supported.

Subjectively speaking, I think the pen latency on this device is near nonexistent and on par with CTL-480 on notes app (iPadOS) and the display is 120hz. Of course, I’ll have to test this after Apple Pencil prediction is added to lazer.

I’m enthusiastic about pen tablets so I used to have the original Surface Pro (a different Wacom, same as S-pens) and currently have a Surface Pro 5 (n-trig) tablet as well and on those devices the pens are not suitable for osu but with some heavy latency correction I think users can still have tons of fun with those devices.

I started playing osu on a pen tablet like Surface Pro (before that, osu!Stream and the original game on nintendo DS) and tapping the notes directly is an extremely intuitive and fun experience (with lower entry barrier) that a lot of new players will prefer if given the choice between desktops.

**Comment by jai-x** [comment]
I don't think predictive movement should be added for touch devices.
Predictive movement itself is used as QoL enhancement to make up for the latenecy of input.
For osu!, as most of the gameplay in itself is a measurement of your input accuracy with regards to a beatmap, having predictive movement seems like an unfair assistance.

**Comment by peppy** [comment]
This is a framework issue. It should definitely be made available as an option.

**Comment by Snowmander** [comment]
The wacom drivers all have in-built prediction to different degree based on versions, and on stable users are allowed to use whatever movement predicting driver. Latency correction has been part of osu! ever since players began using tablets.

Also, latency correction on iPad only compensates for technical limitations, not perform player mind reading magic (which can not exist).

As long as one input stream is ultimately used for judgement, it’s absolutely fine.

**Comment by jai-x** [comment]
Makes more sense when put that way, thanks.

**Comment by swoolcock** [comment]
Agreed that this would be a nice addition to the framework.  It should be fairly trivial to expose touch gesture prediction on iOS, as the "virtual" events are provided as part of each `UITouch` object.
